mktempdir() now supports prefix.

sambitdash opened the pull request:
The libuv is used for this implementation in a platform independent manner. This is for the issue #22922. The `mktemp()` or `tempname()` functions are not affected by this change. 

This may be still acceptable as a module owner can create her own temp directory and then create files inside that. Hence, the importance of prefix on files may not be as significant. 

Edit: fix #22922

rfourquet committed:
re-enable the 0-arg MersenneTwister() constructor

The arguments to `srand(rng, ...)` and `typeof(rng)(...)`
should mirror each-other, in particular
`srand(MersenneTwister(0))` and `MersenneTwister()` should
produce an equivalent object.
Also, `srand(::RandomDevice)` has been added for consistency,
which could be useful in generic code.

vtjnash committed:
revise boundscheck structure

it is much easier if this value gets treated as a normal parameter
as that allows all of the normal control flow logic to apply
rather than require a complete reimplementation of it

JeffBezanson committed:
add `pairs`, and `keys`/`values` for arrays

This allows more uniform treatment of indexed collections.

JeffBezanson committed:
use `pairs` in `findmin` and `findmax`, supporting all indexable collections

return `CartesianIndex` for n-d arrays in findmin, findmax, indmin, indmax

more compact printing of `CartesianIndex`

change sparse `_findr` macro to a function

nalimilan committed:
Rename select* functions to partialsort*

The new name is more explicit, more consistent with sort and with one of the
most commonly used names for this operation (from the C++ stdlib). It also
does not conflict with other meanings e.g. for POSIX sockets and SQL.

nalimilan committed:
Make partialsort!() and partialsortperm!() return a view rather than a copy

Returning a copy (partially) defeats the purpose of these functions, which is
to avoid allocations.

rfourquet committed:
REPL: enable undo in bracketed paste mode

sambitdash committed:
mktempdir() now supports prefix.

The libuv is used for this implementation in a platform independent
manner.

sambitdash committed:
mktempdir() now supports prefix.

The libuv is used for this implementation in a platform independent
manner.

sambitdash committed:
mktempdir() now supports prefix.

The libuv is used for this implementation in a platform independent
manner.

sambitdash committed:
mktempdir() now supports prefix.

The libuv is used for this implementation in a platform independent
manner.

sambitdash committed:
mktempdir() now supports prefix.

The libuv is used for this implementation in a platform independent
manner.

sambitdash committed:
mktempdir() now supports prefix.

The libuv is used for this implementation in a platform independent
manner.

sambitdash committed:
mktempdir() now supports prefix.

The libuv is used for this implementation in a platform independent
manner.

sambitdash commented:
With this #22998 can be closed.

sambitdash committed:
Documentation error fix.

sambitdash committed:
Documentation error fix.

sambitdash committed:
Documentation error fix.

sambitdash committed:
Documentation error fix.

sambitdash committed:
Documentation error fix.

sambitdash committed:
Documentation error fix.

sambitdash committed:
Documentation error fix.

sambitdash committed:
OS specific test anomalies detected.

sambitdash committed:
OS specific test anomalies detected.

sambitdash commented:
Added test cases which show behavioral differences across OS. For example, '*' is not an accepted prefix in Windows which is perfectly permitted in Linux as part of a filename.

sambitdash commented:
Error reported in Mac OS X is completely unrelated. It's an InterruptException which this code does not deal with. The Windows 32-bit case is more like incomplete run and is not a true failure.

JeffBezanson committed:
make `promote` throw an error if no arguments can be changed. fixes #22801.

JeffBezanson committed:
implement #6614, destructuring in formal arguments

rfourquet committed:
REPL: register previous run action as symbol

rfourquet committed:
REPL: add few "region" operations

* add possibility to set the "mark" in the edit area, with C-Space
* the mark and current position (the "point") delimit a region
* C-x C-x exchanges the mark with the point
* M-w is set to copy the region into the kill buffer
* M-W is set to kill the region into the kill buffer
  (in Emacs, the binding for this is C-w, but it's already taken
  by edit_werase)

quinnj committed:
Work to ensure immutable isbits Union fields get set correctly. #23351

rfourquet committed:
REPL: free-up '\0' as a key (sent by Ctrl-Space)

'\0' was used as a sentinel for self-insert (spelled "*" in keymaps),
which made it un-available for Ctrl-Space combo.
We replace here '\0' with an un-assigned sentinel Char value.

rfourquet committed:
REPL: implement a kill ring (fix part of #8447)

After a yank ("^Y"), it's possible to "yank-pop" with
the "M-y" binding, which replaces the just-yanked text
with an older entry from the "kill-ring".

rfourquet committed:
REPL: specify regions with open-close ranges using Pairs

All call sites of `splice_buffer!` were manually substracting 1 to
the last position of the range, indicating that it's more natural
to specify a region with an open-close range.
Therefore, `splice_buffer!` and `edit_replace` (thin wrapper of
`splice_buffer!`) are merged into `edit_splice!`, which takes
a `Region` (a pair of integers) object instead of a range.

sambitdash commented:
Closing as there is not much of activity in past week on this. If you there is interest in this change let me know, I will be happy to reopen.

StefanKarpinski commented:
Seems a bit premature to close this. @nalimilan, you seemed engaged on the original issue here. Can you review this and see if it addresses the problem? Frankly, I don't have enough context to understand what problem this is solving, but that doesn't mean it's not a good idea.

sambitdash committed:
A test error previously not reported fixed.

sambitdash committed:
A test error previously not reported fixed.

sambitdash commented:
PR#23259  affected the last test errors. Now fixed.

jw3126 committed:
Make printing of big(NaN) consistent (#23382)

vtjnash committed:
correctly handle union-splitting in codegen type allocation

JeffBezanson committed:
allocate fresh bindings on each iteration of `while` loops

also remove the outer scope "wrapper" from `while` loops, similar to #23387

StefanKarpinski committed:
Make --quiet and --banner independent (fix #23380)

sambitdash commented:
Travis failure seems unrelated.

StefanKarpinski commented:
@vtjnash has said he can take a look at this in a few days. Sorry for the delay, but this is a useful change.

quinnj committed:
Add tests for structs with isbits Union fields.

nalimilan commented on base/file.jl:
```diff
@@ -356,12 +339,33 @@ is an open file object for this path.
+    template = prefix*"XXXXXX"
+    tpath = joinpath(parent, template)
+
+    req = Libc.malloc(_sizeof_uv_fs)
```
I see you use the same approach as for `readlink`, but `readdir` uses a different strategy, by using `zeros`, which allows skipping the `try... finally`. @vtjnash Which approach is recommended?

sambitdash replied:
I guess I now get the approach here. In `jl_uv.c` all request structures are allocated in stack (declared as local var) and passed on to the function in the c-layer. The cleanup is carried out on the address to the request like &req. In the Julia layer you cannot have a stack variable so `zero` is a means to create a Julia UInt8 array which will me memory managed in the Julia layer. Hence, no special action on a throw and clean up will be called on this object after the request object is used in the `ccall`. The approach is a bit hard to understand initially.  Are `try`...`catch` overheads very significant? I can change the code to the `zero` approach.

nalimilan replied:
I would think the simplest approach is the best one, but others must know better.

nalimilan commented on test/file.jl:
```diff
@@ -1156,3 +1156,46 @@ if !Sys.iswindows()
+        @test_throws Base.UVError mktempdir(; prefix="*")
+        @test_throws Base.UVError mktempdir(; prefix="cdcdccd/")
+    else
+        # '/' is accepted in a prefix but affects the overall path and permissions.
```
That comment sounds misleading, since `/` is not accepted by the next line.

sambitdash replied:
The whole behavior is very messy. Here are the full set of samples. I think we should just not accept '/' in prefix. Some samples:

```
julia> mktempdir("/";prefix="tmp/")
"/tmp/m12rZm"

julia> mktempdir("/tmp";prefix="/")
ERROR: mktempdir: permission denied (EACCES)
Stacktrace:
 [1] uv_error at ./libuv.jl:68 [inlined]
 [2] #mktempdir#12(::String, ::Function, ::String) at ./file.jl:359
 [3] (::Base.Filesystem.#kw##mktempdir)(::Array{Any,1}, ::Base.Filesystem.#mktempdir, ::String) at ./<missing>:0
```

nalimilan replied:
I'm not too worried about this, as these behaviors make sense, but differences between OSes are annoying. What happens with `\` on Windows?

nalimilan commented on test/file.jl:
```diff
@@ -1156,3 +1156,46 @@ if !Sys.iswindows()
+    # Behavioral differences across OS types
+    if Sys.iswindows()
+        @test_throws Base.UVError mktempdir(; prefix="*")
+        @test_throws Base.UVError mktempdir(; prefix="cdcdccd/")
```
Does `/` without any characters before it also throw? What about `\` or `C:\`?

sambitdash replied:
Give me a couple of days. My windows VM is needs some fixing before I can test this. I am a native Linux programmer :-).

nalimilan commented on test/file.jl:
```diff
@@ -1156,3 +1156,46 @@ if !Sys.iswindows()
+    end
+
+    # Unicode test
+    tst_prefix="\u2200x\u2203y"
```
Directly use the Unicode chars, that's clearer.

nalimilan reviewed (commented):
Thanks! Looks mostly good, though I'd like a deeper review of libuv parts by somebody familiar with that code. Please improve the commit message to explain that you switch to the new `uv_fs_mkdtemp` function, and which libuv version it was introduced in.

jw3126 committed:
More compact BigFloat printing (#23382)

sambitdash commented on base/file.jl:
```diff
@@ -356,12 +339,33 @@ is an open file object for this path.
+    req = Libc.malloc(_sizeof_uv_fs)
+    try
+        ret = ccall(:uv_fs_mkdtemp, Int32,
+            (Ptr{Void}, Ptr{Void}, Cstring, Ptr{Void}),
```
I followed the same indentation as `readlink`, although my personal preference would be to align to ccall.

vtjnash committed:
correctly handle union-store-splitting in "is" and "object_id"

kshyatt committed:
Doc checkout methods and type (#23410)

* Doc checkout methods and type

rfourquet committed:
REPL: add tests for the kill ring

rfourquet committed:
REPL: fix newline indentation when point before 1st non-space (#23402)

rfourquet committed:
fix prompt-indentation when bracket pasting (#23406)

Cf. problem reported there:
https://github.com/JuliaLang/julia/pull/22948#issuecomment-324093541
Bracket pasting was not removing the indentation corresponding
to the prompt (i.e. 7 spaces for "julia> ").

musm committed:
Int constructor instead of div for Int returns in gmp (#23329)

vtjnash committed:
fix codegen bugs, simplify type representations

anonymize all types, remove special representation for Ptr, fix Constant
generation for types containing split-union fields

Sacha0 committed:
deprecate expm in favor of exp (#23233)

* Deprecate expm in favor of exp.

* Remove reference to expm in the linalg docs

JeffBezanson committed:
restrict `Bool` `*` methods to `AbstractFloat`, avoiding some ambiguities

part of #19168

Commit:
Doc-strings for some internals. (#23333)

* REPL-documentation for more internals.

* Copy edit 'REPL-documentation for more internals.'

* New section "Special Types"

kshyatt committed:
Document merge base and GitAnnotated

kshyatt committed:
Add docs for transact

musm committed:
Unify the method name to get library versions (#23323)

kshyatt committed:
Add more information for CherrypickOptions

ararslan committed:
Preserve the input element type in unique (#23208)

Previously the element type of the output was the smallest type that
would fit the union of the input's individual element types. Now the
output has an identical element type to the input. Fixes #22696.

rfourquet committed:
REPL: implement Alt-{u,c,l} to change the case of the next word (#23379)

Fixes part of #8447.

musm committed:
Misc remove `&x` ccall  syntax (#23301)

musm committed:
Use iszero and isone in MPFR (#23322)

maleadt committed:
Merge pull request #22594 from JuliaLang/tb/optpasses_tm

Make addOptimizationPasses target-agnostic.

0x47 committed:
Disable globbing for curl download (#23254)

iglpdc committed:
[DOC] Document Void type (#23324)

Fixes #22806

musm committed:
Use refs instead of Vector in Lapack (#23419)

* Use refs instead of Vector in Lapack

* Use uninitialized refs

* Manual constructors

* Simply

JeffBezanson committed:
fix #23430, counter overflow in subtyping

kshyatt committed:
Add docs for GitRevWalker and push_head! (#22654)

* Add docs for GitRevWalker and push_head!

nalimilan committed:
Fix docstrings for partialsortperm() and partialsortperm!()

partialsort(x, k::Integer), partialsortperm(x, k::Integer) and partialsortperm!(x, k::Integer)
all return a single value. Also add missing backticks.

omus committed:
Merge pull request #23412 from JuliaLang/cv/env-get

Add method for `get(f::Function, ::EnvHash, key)`

StefanKarpinski committed:
Merge pull request #23399 from JuliaLang/sk/quieter

Make --quiet and --banner independent (fix #23380)

JeffBezanson committed:
Merge pull request #23411 from JuliaLang/jb/parseable

parseable printing of all type names and TypeVars

kshyatt committed:
Doc rebase stuff (#23432)

* Doc rebase stuff

* fix version stuff

picked "newer than or equal to" for consistency with existing docstrings

kshyatt committed:
Merge pull request #23437 from JuliaLang/ksh/docmerge2

Document merge base and GitAnnotated

StefanKarpinski committed:
fix PR #23399 accidentally turning off banner

kshyatt committed:
Merge pull request #23439 from JuliaLang/ksh/doccpo

Add more information for CherrypickOptions

kshyatt committed:
Merge pull request #23438 from JuliaLang/ksh/doctransact

Add docs for transact

kshyatt committed:
Docs for libgit2 utils

kshyatt committed:
Docs for a bunch of git blame stuff

kshyatt committed:
Doc status opts and fix a typo

omus committed:
Return GitError from credential_loop

timholy committed:
Describe how to hack Base more easily with Revise.jl

[ci skip]

visr committed:
fix printing expressions like Ref{<:Real}

fixes #23457

Previously they were not explicitly handled by show_unquoted,
ending up being printed as Expr. This was not parsable back
to their original meaning.

musm committed:
Remove `&x` convention in ccall in gmp (#23289)

StefanKarpinski committed:
Merge pull request #23453 from JuliaLang/sk/banner

fix PR #23399 accidentally turning off banner

StefanKarpinski committed:
Merge pull request #23444 from 0x47/curl_globbing

Disable globbing for curl download (#23254)

musm committed:
Remove `&x` convention in ccall in mpfr (#23288)

omus committed:
challenge_prompt now takes an expression

Switch from strings to expressions for readability and to improve
support for generating tests.

yuyichao committed:
Make sure error from `Expr(:new)` is not optimized away. (#23353)

* Fix effect_free for `Expr(:new)`

  Check if the type being constructed is leaftype and check if all fields are correctly typed.
  The expression could throw an error otherwise.

* Fix functions used by interpreter and codegen to check field types

* Fix (remove) assumption of `Expr(:new)` being effect free in lowering

bicycle1885 committed:
fix a typo in a docstring (#23466)

Keno committed:
Update conditional documentation example

It confuses femtocleaner (https://github.com/JuliaLang/julia/pull/23440) and it's probably
better to not reference a specific julia version in the documentation anyway.

vtjnash committed:
inference: use TypedSlot to mark intermediate load types

previously, we widened the load type too soon, discarding potentially valuable information earlier than necessary

vtjnash committed:
inference: ensure NewvarNode side-effect is handled

vtjnash committed:
update libuv (#23326)

fixes UV_HANDLE_EMULATE_IOCP handling for Windows pipes

close #11727

timholy committed:
Merge pull request #23446 from JuliaLang/teh/revise

Describe how to hack Base more easily with Revise.jl

kshyatt committed:
Merge pull request #23455 from JuliaLang/ksh/docblame

Docs for a bunch of git blame stuff

kshyatt committed:
Merge pull request #23458 from JuliaLang/ksh/docstatus

Doc status opts and fix a typo

kshyatt committed:
Merge pull request #23454 from JuliaLang/ksh/docutils

Docs for libgit2 utils

Keno committed:
Merge pull request #23467 from JuliaLang/kf/docexample

Update conditional documentation example

kshyatt committed:
Even more docs (#23420)

* Even more docs

kshyatt committed:
Doc merge_analysis and some enums (#23435)

* Doc merge_analysis and some enums

nalimilan committed:
Merge pull request #23051 from JuliaLang/nl/select

Rename select* functions to partialsort* and various related fixes

ararslan committed:
Add docs for git oid methods (#23460)

kshyatt committed:
Doc SigStruct and DescribeOptions

JeffBezanson committed:
Merge pull request #23464 from visr/printsugar

fix printing expressions like Ref{<:Real}

JeffBezanson committed:
fix bug in showing `UnionAll` introduced by #23411

JeffBezanson committed:
fix bug in printing generator expressions with n-d ranges

JeffBezanson committed:
Merge pull request #23479 from JuliaLang/jb/showUnionAll

fix bug in showing `UnionAll` introduced by #23411

JeffBezanson committed:
Merge pull request #23480 from JuliaLang/jb/showgenerator

fix bug in printing generator expressions with n-d ranges

JeffBezanson committed:
Merge pull request #23447 from JuliaLang/jb/fix23430

fix #23430, counter overflow in subtyping

vtjnash committed:
Merge pull request #23470 from JuliaLang/jn/var-infer

better constant variable inference

JeffBezanson committed:
deprecate `catch <expression>`. addresses #19987

femtocleaner[bot] committed:
Fix deprecations

JeffBezanson committed:
fix #23107, ensure `Array` constructor always makes new arrays

JeffBezanson committed:
Merge pull request #23433 from JuliaLang/jb/boolmul

restrict special `*` methods for `Bool`, which tend to cause ambiguities

vtjnash committed:
fix missing propagate_inbounds from ordering comparison

omus committed:
Fix GIT_CONFIG docstring

kshyatt committed:
Make describe docs more accurate/informative (#23477)

* Make describe docs more accurate/informative

kshyatt committed:
Merge pull request #23476 from JuliaLang/ksh/docdoc

Doc SigStruct and DescribeOptions

saharjs committed:
Update README.md (#21326)

* Update README.md

* updates

xgdgsc committed:
rm ubuntu nightly ppa as not actively maintained (#23484)

* rm ubuntu nightly ppa as not actively maintained

* rm list of outdated repos

quinnj committed:
Add missing datatype_module method for UnionAll types

fredrikekre committed:
fix some convert methods that depwarned (#23494)

dhoegh committed:
Fix 22577. Add string macros to REPLCompletion. (#23119)

* Fix 22577. Add string macros to REPLCompletion.

* Add cmd macros to REPLCompletion.

* retrigger ci, update to a testset

* unicode tests.

* also complete the unicode char

andreasnoack committed:
Deprecate countnz in favor of using count(predicate, x) (#23485)

Define missing specialization of count(pred, x) for sparse arrays

vtjnash committed:
test-codegen: make jl_dump_compiles test more reliable (#23486)

This was a flaky test, and it's not entirely clear what it was testing.

Common failure was:
        FAILURE
    Error in testset codegen:
    Test Failed
      Expression: tempty == true
       Evaluated: false == true
    ERROR: LoadError: Test run finished with errors
    while loading C:\projects\julia\julia-a661736be5\share\julia\test\runtests.jl, in expression starting on line 29
    Command exited with code 1

mbauman committed:
Add propagate_inbounds for LogicalIndex iteration

This optimization had been occurring before, but still isn't entirely valid

Keno committed:
Merge pull request #23440 from JuliaLang/fbot/deps

Fix deprecations

tuckermcclure committed:
Added notes on closing a library. (#23489)

* Added notes on closing a library.

Fixes 23459

* shorten line lengths

* "can"/"could" consistency.

kshyatt committed:
Doc warn and warn_with (#23441)

* Doc warn and warn_with

JeffBezanson committed:
fix GC rooting of arguments in cfunction `_gfthunk` wrappers

JeffBezanson committed:
fix `promote_rule` for `Bidiagonal` and `Tridiagonal`

JeffBezanson committed:
Merge pull request #23397 from JuliaLang/jb/whilescope

RFC: allocate fresh bindings on each iteration of `while` loops

kshyatt committed:
Finish docs for GitTree (#23451)

* Finish docs for GitTree

musm committed:
Remove global const  in init in cholmod (#23449)

* Remove global const  in init in cholmod

* Make refs and commonStruct a const

* Remove common function

* add back removed space [ci skip]

iblislin committed:
test/file: try to fix freezing of test_22566 (#23001)

- wrap the long single test scripts

- using `$(repr(fn))` for inserting temp file name

- use STDIN to synchronize fifo open calls for each end

kshyatt committed:
More docs for merge! (#23495)

* More docs for merge!

JeffBezanson committed:
Merge pull request #23501 from JuliaLang/jb/gfthunk_roots

fix GC rooting of arguments in cfunction `_gfthunk` wrappers

JeffBezanson committed:
Merge pull request #23487 from JuliaLang/jb/catchexpr

deprecate `catch <expression>`

JeffBezanson committed:
Merge pull request #23503 from JuliaLang/jb/bidiag_promote

fix `promote_rule` for `Bidiagonal` and `Tridiagonal`

martinholters committed:
Fix `countnz` deprecation (#23509)

* `export foo` -> `export countnz`
* `depwarn(..., :depwarn)` -> `depwarn(..., :countnz)`

yuyichao committed:
Make partial_inst a global root. (#23507)

Types in this list might not be rooted anywhere else.

mbauman committed:
Fix off-by-one error in string reallocation (#23493)

and ensure that array to string conversions do not cross GC size boundaries

omus committed:
Fix expanduser with empty string (#23506)

mbauman committed:
Completely remove partial linear indexing (#21750)

This removes the partial linear indexing deprecation and implents the new behavior: Linear indexing now only takes effect when there is exactly one non-cartesian index.

mbauman committed:
Merge pull request #22826 from JuliaLang/jn/boundscheck-revised

RFC: make boundscheck a value

JeffBezanson committed:
Merge pull request #23490 from JuliaLang/jb/Arrayctor

fix #23107, ensure `Array` constructor always makes new arrays

fredrikekre committed:
fix promotion of diagonal integer matrix to non-integer power (#23510)

mbauman committed:
Document boundscheck macro

Fix #20469. [ci skip]

kshyatt committed:
Doc statusentry and add xrefs

mbauman committed:
Add test for issue #20469

JeffBezanson committed:
Merge pull request #23337 from JuliaLang/jb/arg_destructuring

RFC: implement #6614, destructuring in formal arguments

JeffBezanson committed:
Merge pull request #23491 from JuliaLang/jb/noncircular_promote

RFC: make `promote` throw an error if no arguments can be changed

tkelman committed:
Update to latest curl, mbedtls, and pcre (#23498)

* Update to latest curl, mbedtls, and pcre

* reason for the pcre patch was solved differently upstream

fredrikekre committed:
move mathematical constants to Base.MathConstants (#23427)

- only export pi, π and ℯ from Base
 - remove eu as alias for ℯ
 - add \euler as tab completion for ℯ

kshyatt committed:
Merge pull request #23517 from JuliaLang/ksh/docstatus

Doc statusentry and add xrefs

KlausC committed:
make findmin/findmax behavior match min/max (fix #23209) (#23227)

ararslan committed:
Deprecate rol and ror in favor of circshift methods (#23404)

Sacha0 committed:
Deprecate logm in favor of log. (#23505)

fredrikekre committed:
fix documentation for IOBuffer, close #23398 (#23514)

Sacha0 committed:
Deprecate sqrtm in favor of sqrt. (#23504)

kshyatt committed:
Add docs for internal commit method

fredrikekre committed:
clarification about ℯ (#23534)

kshyatt committed:
Merge pull request #23538 from JuliaLang/ksh/doccommit

Add docs for internal commit method

JeffBezanson committed:
parse `let` the same as `for`. part of #21774

kshyatt committed:
Add field docs for Diffopts

JeffBezanson committed:
deprecate `&x` in ccall. fixes #6080

Commit:
Cleaned up default descriptions in reflection.jl (#23543)

* Cleaned up default descriptions in reflection.jl

Redundant default descriptions were removed. In addtion, square brackets to denote optional arguments were replaced by default syntax:

code_native([io], f, types, [syntax]) -> code_native(io=STDOUT, f, types, syntax=att)

* Use `[io=STDOUT,]` to show default argument

Commit:
update rationalize docs (#23541)

Removed description of default for T. Was redundant with signature

rfourquet committed:
fix StackOverflowError in symdiff(::IntSet, ::NotInteger)

StephenVavasis committed:
more about type annotation (#23180)

* more about type annotation

Type annotation not recommended for types constructed at run-time

rfourquet committed:
fix IntSet tests invalid since #23138

Now IntSet have the length of their underlying BitVector be
a multiple of 64, so resizing them to something else breaks
this invariant.

JeffBezanson committed:
fix #23558, bug in recursive let-bound functions

fredrikekre committed:
faster sparse(::Diagonal) (#23536)

yuyichao committed:
Makes sure assertion respects JL_NDEBUG in C++ code

This generalizes `fix_llvm_assert.h` to `julia_assert.h` and replace (almost) all uses of
`cassert` and `assert.h` with `julia_assert.h`.
Changes the rule for including `julia_assert.h` so that includes of `assert.h` from external
headers (including `libsupport`) won't accidientally disables assertion.
Without this patch, the `assert.h` included in `libsupport.h` (in `julia.h`) may disable
assertions in our c++ code.

Keno committed:
Add gc use intrinisic

Now that our gc root placement pass is significantly more aggressive about
dropping roots, we need to be a bit more careful about our use of pointer,
etc. This adds a low-level intrinsic for annotating gc uses to keep objects
alive even if they would be otherwise unreferenced.

As an initial use case, we get rid of a number of uses of `pointer` in string,
but creating a new `unsafe_load` that keeps the string alive.

andreasnoack committed:
Use min(size(A)...) instead of max(size(A)...) when computing the numrical rank (#23523)

Fixes #23386

Keno committed:
Don't ignore memcpy in GC root placement

JeffBezanson committed:
Merge pull request #23559 from JuliaLang/rf/fix-intset-test

fix IntSet tests invalid since #23138

Keno committed:
Merge pull request #23563 from JuliaLang/yyc/build/assert

Makes sure assertion respects JL_NDEBUG in C++ code

Keno committed:
Remove some leagcy gcroot codegen code

The LLVM pass now figures this out automatically, so we no longer
need to put it explicitly into the code. There's more cleanup
to be done here, but I want to take this one step at a time,
since this legacy code may have been hiding issues in the GC
root placement pass.

rfourquet committed:
Merge pull request #23377 from JuliaLang/rf/REPL-killring

REPL: implement a kill ring

rfourquet committed:
Merge pull request #21909 from JuliaLang/rf/MT-ctor

re-enable the 0-arg MersenneTwister() constructor

Keno committed:
Merge pull request #23564 from JuliaLang/kf/gcmemcpy

Don't ignore memcpy in GC root placement

Keno committed:
Merge pull request #23562 from JuliaLang/kf/gcuse

Add gc use intrinisic

JeffBezanson committed:
Merge pull request #23557 from JuliaLang/rf/IntSet-symdiff-overflow

fix StackOverflowError in symdiff(::IntSet, ::NotInteger)

JeffBezanson committed:
Merge pull request #23561 from JuliaLang/jb/fix23558

fix #23558, bug in recursive let-bound functions

maleadt committed:
Add a codegen param for preferring specsig.

Keno committed:
Merge pull request #23568 from JuliaLang/kf/legacygc

Remove some legacy gcroot codegen code

Keno committed:
Untangle legacy .gcroot field

When the new gc root placement pass was introduced, the .gcroot field
lost its original meaning and became mostly superfluous. However, the
codegen code for unions still used it to get a referenced to the boxed
part of the split union. Clean that up by introducing an explicit field
for that purpose and removing the .gcroot field.

Keno committed:
Untangle legacy .gcroot field

When the new gc root placement pass was introduced, the .gcroot field
lost its original meaning and became mostly superfluous. However, the
codegen code for unions still used it to get a referenced to the boxed
part of the split union. Clean that up by introducing an explicit field
for that purpose and removing the .gcroot field.

kshyatt committed:
Merge pull request #23554 from JuliaLang/ksh/docdiffopts

Add field docs for Diffopts

rfourquet committed:
REPL: enable "redo" after "undo"

omus committed:
Correct indentation for LibGit2.Consts

omus committed:
Fix GIT_OPT indentation

StefanKarpinski committed:
Merge pull request #23388 from jw3126/big

Make printing of big(NaN) consistent #23382

quinnj committed:
Fix #23567. In general, we need to be careful when we start branching on isbits(eltype(::Array)) because of Union isbits optimizations because isbits Unions don't return isbits == true.

quinnj committed:
Remove incorrect code from jl_array_ptr_copy for isbits Union arrays

quinnj committed:
Update unsafe_convert for isbits Unions according to recommendation from yuyichao.

fredrikekre committed:
Unify function references in docs (#23535)

* Unify function references by removing misleading ()

* review fix

JeffBezanson committed:
Merge pull request #23555 from JuliaLang/jb/ampersand

deprecate `&x` in ccall. fixes #6080

ararslan committed:
Add a trailing newline to the & deprecation message

ararslan committed:
Fix deprecated & syntax in MPFR ccalls

maleadt committed:
Embed codegen hooks in the params object.

Using pointer_objref is unsafe for the GC.

maleadt committed:
Remove cgparams that disallow language features.

We realistically rely on LLVM optimization passes anyway.

Keno committed:
`unsafe_load` -> `@inbounds codeunit`

Addresses post-commit review in #23562.

quinnj committed:
Probably due to #23397; ensure the istart variable is declared outside the while loop

quinnj committed:
Merge pull request #23577 from JuliaLang/jq/bitsunionarrays

Fix #23567

quinnj committed:
Merge pull request #23496 from JuliaLang/jq/refl

Add missing datatype_module method for UnionAll types

ararslan committed:
Merge pull request #23580 from JuliaLang/aa/ref-dep-fix

Fix & deprecation message and two uses of that syntax

halleysfifthinc committed:
Switch Cygwin download links to https (#23578)

fredrikekre committed:
remove mention of helpdb/Base.jl in CONTRIBUTING.md (#23587)

Commit:
REPL: Add ^P and ^N to modal prefix searching (#23319)

* Add ^P and ^N to modal prefix searching
* Add meta-P, meta-N as synonyms for page up/page down in REPL

HarrisonGrodin committed:
Add REPL completion for tilde paths (#23475)

Overrides path completions if a user expansion is available. For
example, `~/Docu` will autocomplete to `/home/user/Docu` after one
<TAB>, and then to `/home/user/Documents` after a second <TAB>.

rfourquet committed:
remove vestigial Char in rand(::UnitRange{<:Union{..., Char}) (#23545)

Cf. #13600 and https://github.com/JuliaLang/julia/pull/22752/files#r128026174.
Also, changing `Union{Signed,Unsigned,BigInt,Bool}` to `Integer`, as this
is equivalent in practice.

quinnj committed:
Merge pull request #23584 from JuliaLang/jq/cartesian

Fix while loop variable issue in cartesian.jl

rfourquet committed:
Merge pull request #9596 from srp/repl-undo

JeffBezanson committed:
Merge pull request #22907 from JuliaLang/jb/keyvalue

make indexable collection API more uniform (keys, values, pairs)

JeffBezanson committed:
fix #15723, `findfirst` on strings

Makes findfirst, findlast, findnext, and findprev more generic
by using endof, nextind, and prevind. Also adds a `nextind` method
for arrays and `CartesianIndex`.

vtjnash committed:
add error message for attempting to capture a local variable as a closure type signature (#23527)

fix #18730

Keno committed:
Merge pull request #23582 from JuliaLang/kf/codeunit

`unsafe_load` -> `@inbounds codeunit`

Keno committed:
Merge pull request #23571 from JuliaLang/kf/untanglegcroot

Untangle legacy .gcroot field

maleadt committed:
Merge pull request #23581 from JuliaLang/tb/inflight

Fix and improve codegen params

JeffBezanson committed:
Merge pull request #23553 from JuliaLang/jb/parselet

parse `let` the same as `for`. part of #21774

rfourquet committed:
LineEdit: add convenient default mode in `state` function

JeffBezanson committed:
Merge pull request #23592 from JuliaLang/jb/fix15723

fix #15723, `findfirst` on strings

omus committed:
Exclude starting slash from path in URL_REGEX (#23574)

Keeps URL parsing consistent between the scp-like syntax and the scheme
based parsing.

omus committed:
Display STDOUT from Pkg.build

jrevels committed:
fix boundscheck value propagation codegen bug (#23595)

vtjnash committed:
inference: leaftype Type{typeof(Union{})} is ambiguous

fix #23413

vtjnash committed:
invoke should raise MethodError for ambiguous cases

fix #18095
fix #22988 (by improving error message)

vtjnash committed:
fieldtype validation: disallow `Vararg` as a field type

fix #17951

JeffBezanson committed:
remove uses of `importall` in Base

omus committed:
Merge pull request #23572 from JuliaLang/cv/libgit2-consts

Fix indentation with LibGit2.Consts

JeffBezanson committed:
Merge pull request #23602 from JuliaLang/jn/23413

inference: leaftype Type{typeof(Union{})} is ambiguous

quinnj committed:
Speed up isbitsunion and bitsunionsize functions

rfourquet committed:
REPL undo: fix edit_kill_region and add more tests

vchuravy committed:
Merge pull request #23601 from JuliaLang/cv/build-stdout

Display STDOUT from Pkg.build

rfourquet committed:
Merge pull request #23590 from JuliaLang/rf/repl-redo

REPL: enable "redo" after "undo"

JeffBezanson committed:
Merge pull request #23603 from JuliaLang/jn/18095

invoke should raise MethodError for ambiguous cases

JeffBezanson committed:
Merge pull request #23604 from JuliaLang/jn/17951

fieldtype validation: disallow `Vararg` as a field type

rfourquet committed:
make BigInt <: Signed (#23473)

mbauman committed:
Merge pull request #23516 from JuliaLang/mb/docboundscheck

Document boundscheck macro

quinnj committed:
Merge pull request #23586 from JuliaLang/jq/bitsunionsperf

Speed up isbitsunion and bitsunionsize functions

quinnj committed:
Merge pull request #23367 from JuliaLang/jq/bitsunions

Ensure immutable isbits Union fields get set correctly

ararslan committed:
Add a README note referencing FreeBSD Pkg segfaults (#23613)

omus committed:
Merge pull request #23573 from JuliaLang/cv/libgit2-helper-error

Refactor LibGit2 credential_loop tests

ViralBShah committed:
Add GPU support to the README. (#23623)

* Add GPU support to the README.

* Add PTX backend note.

* External is a better word for the PTX support, I think.

kshyatt committed:
[RFC] Clean up test/tridiag (#23609)

* Clean up test/tridiag

JeffBezanson committed:
Merge pull request #23608 from JuliaLang/jb/no_importall

remove uses of `importall` in Base

rfourquet committed:
fix #19182: convert([Un]Signed, x::BigInt) (#23474)

rfourquet committed:
REPL: more deleted words pushed into the kill ring (#23593)

The two ways of deleting the previous word, deleting
the next word and clearing the space now hook into the
kill ring. Implements item 2 of #8447.

kshyatt committed:
Remove string "Array" funnybusiness (#23583)

* Remove string "Array" funnybusiness

* Remove last string things and fix comments

* Re-add missing test

sambitdash committed:
Merge branch 'I22922' of https://github.com/sambitdash/julia into I22922

sambitdash committed:
Merge branch 'I22922' of https://github.com/sambitdash/julia into I22922

sambitdash commented:
closing and resubmitting with rebasing.

fredrikekre commented:
You can rebase and force-push to the same branch and keep the same PR.

sambitdash commented:
@fredrikekre  The forced push detached the PR and could not reopen it. The new PR on this is #23634.

fredrikekre commented:
Yea, I think you need to reopen before force-pushing in such a case.
